src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java

```diff
@@ -184,30 +233,109 @@ private static IntervalList padIntervals(final List<SimpleInterval> inputInterva
+            for (byte b : bases) {
+                if (Nucleotide.decode(b) != Nucleotide.N) {
+                    return true;
+                }
             }
-        }
-        return bins;
+            return false;
+        });
     }
 
     @Override
     public void traverse() {}  // no traversal for this tool
+
+    private static class IntervalSpliterator implements Spliterator<Interval> {
+
+        private int binStart;
+        private final Interval interval;
+        private final int binLength;
+        private final boolean grid;
+        private final int minBinLength;
+
+        IntervalSpliterator(final Interval interval,
+                            final int binLength,
+                            final int minBinLength,
+                            final boolean grid) {
+            this.binStart = interval.getStart();
+            this.interval = interval;
+            this.binLength = binLength;
+            this.minBinLength = minBinLength;
+            this.grid = grid;
+        }
+
+        @Override
+        public boolean tryAdvance(final Consumer<? super Interval> action) {
+             while (true) {
+                 int gridOffset;
+                 // no enough bases left for a minimum bin?
+                 if (binStart > interval.getEnd() - minBinLength + 1) {
+                    return false;
+                    // no gridded bins
+                 } else if (!grid || (gridOffset = binStart % binLength) == 1) {
+                    int binEnd = FastMath.min(binStart + binLength - 1, interval.getEnd());
+                    action.accept(new Interval(interval.getContig(), binStart, binEnd));
+                    binStart = binEnd + 1;
+                    return true;
+                    // gridded bins:
+                 } else { // grid == true && not on grid.
+                    // just 1bp before the next grid start!
+                    if (gridOffset == 0) {
+                        if (minBinLength <= 1) {
+                            action.accept(new Interval(interval.getContig(), binStart, binStart++));
+                            return true;
+                        } else {
+                            binStart++;
+                            // continue:
+                        }
+                    } else {
+                        int binEnd = FastMath.min(binStart + binLength - gridOffset, interval.getEnd());
+                        if (minBinLength <= binEnd - binStart + 1) { // gridOffset > 1
+                            action.accept(new Interval(interval.getContig(), binStart, binEnd));
+                            binStart = binEnd + 1;
+                            return true;
+                        } else { // we skip to next grid start and we try again:
+                            binStart = binEnd + 1;
+                            // continue;
```

Comment:
semicolon here?  (I would just get rid of colons and semicolons in all comments.  This is an especially confusing one, because it could be taken as a commented out continue statement!)